homeassistant/components/template/weather.py

```diff
@@ -327,4 +375,71 @@ async def async_added_to_hass(self) -> None:
                 "_forecast",
                 self._forecast_template,
             )
+
+        if self._forecast_daily_template:
+            self.add_template_attribute(
+                "_forecast_daily",
+                self._forecast_daily_template,
+                on_update=partial(self._update_forecast, "daily"),
+                validator=partial(self._validate_forecast, "daily"),
+            )
+        if self._forecast_hourly_template:
+            self.add_template_attribute(
+                "_forecast_hourly",
+                self._forecast_hourly_template,
+                on_update=partial(self._update_forecast, "hourly"),
+                validator=partial(self._validate_forecast, "hourly"),
+            )
+        if self._forecast_twice_daily_template:
+            self.add_template_attribute(
+                "_forecast_twice_daily",
+                self._forecast_twice_daily_template,
+                on_update=partial(self._update_forecast, "twice_daily"),
+                validator=partial(self._validate_forecast, "twice_daily"),
+            )
+
         await super().async_added_to_hass()
+
+    @callback
+    def _update_forecast(
+        self,
+        forecast_type: Literal["daily", "hourly", "twice_daily"],
+        result: list[Forecast] | TemplateError,
+    ) -> None:
+        """Save template result and trigger forecast listener."""
+        attr_result = None if isinstance(result, TemplateError) else result
+        setattr(self, f"_forecast_{forecast_type}", attr_result)
+        self.hass.create_task(self.async_update_listeners([forecast_type]))
+
+    @callback
+    def _validate_forecast(
+        self,
+        forecast_type: Literal["daily", "hourly", "twice_daily"],
+        result: list[Forecast] | TemplateError,
+    ) -> list[Forecast] | None:
+        """Validate the forecasts."""
+        if result is None or isinstance(result, TemplateError):
+            return None
```

Comment:
If the template rendered to None, it's fine to return None, but if we got a TemplateError, shouldn't we raise?
Also, we should check the result is a list and raise if it isn't.

Reply:
The template will take care of that and skip the validator and go directly to `_update_forecast()`.
Didn't we say to always return an empty list even if there is something wrong with the forecast?